ARROW-13403: [R] Update developing.Rmd vignette (#10930)
apache/arrow: r/vignettes/developing.Rmd

---
title: "Arrow R Developer Guide"
output: 
  html_document:
      toc: true

>       toc_depth: 2

Reviewer: I'm curious if there's a general movement away from `rmarkdown::html_vignette`? I've always used `rmarkdown::html_vignette` because I thought it was streamlined for inclusion in the package (and pkgdown overrides this stuff anyway for the web docs).

Author: Oops, I chucked this in so it made it easier to preview the list of contents when I was rendering this as RMarkdown

vignette: >
  %\VignetteIndexEntry{Arrow R Developer Guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

> 

Reviewer: Super nit picky: is this new line needed?

Author: I'll remove

set -x
```

If you're looking to contribute to arrow, this vignette can help you set up a development environment that will enable you to write code and run tests locally. It outlines:

* how to build the components that make up the Arrow project and R package
* some common troubleshooting and workflows that developers use

This document is intended only for developers of Apache Arrow or the Arrow R package. Users of the package in R do not need to do any of this setup. If you're looking for how to install Arrow, see [the instructions in the readme](https://arrow.apache.org/docs/r/#installation).

For clarity, a quick note on terminology used in this vignette:

* "Apache Arrow" or "Arrow" - the Apache Arrow project, including implementations in different languages
* "libarrow" or "the Arrow C++ library" - Arrow's C++ library (N.B. this term might not be used in documentation in other parts of the Arrow project)
* "arrow" or "the Arrow R package" - the R package

> * `arrow` - a directory into which the Arrow GitHub repo has been checked out

Author: Good point - whilst writing this down was useful while I was working on the vignette, now that terminology is consistent, it's a little redundant.

I'm not 100% on board with the info as a footnote as I think it's important to have it up front still, so what I think will satisfy all requirements here is just explain what `libarrow` is, in parentheses, after the first mention of the term, and then leave everything else to be implicit.

The main point of doing this was the libarrow vs. arrow package distinction, so I think we've ticked those boxes now - though happy to adjust again if you think further changes are needed.

brew install apache-arrow --HEAD
```

### Windows and Linux

On Windows and Linux, you can download a .zip file with the arrow dependencies from the
nightly repository.

To see what nightlies are available, you can use Arrow's (or any other S3 client's) S3 listing functionality to see what is in the bucket `s3://arrow-r-nightly/libarrow/bin`:

```
nightly <- s3_bucket("arrow-r-nightly")
nightly$ls("libarrow/bin")
```
Version numbers in that repository correspond to dates.

#### Windows

Windows users then can set the `RWINLIB_LOCAL` environment variable to point to the zip file containing the arrow dependencies before installing the arrow R package. 

#### Linux

> On Linux, you'll need to create a `libarrow` directory inside the R package directory and unzip the zip file containing the arrow dependencies into it. 

Reviewer: ```suggestion
On Linux, you'll need to create a `libarrow` directory inside the R package directory and unzip the zip file containing the compiled arrow binary files into it. 
```

Reviewer: It does include some dependencies (and technically libarrow *is* a dependency of the R pacakge), but I think this is a clearer description of what we are doing here.

Author: Thanks!

## R and C++

If you need to alter both the Arrow C++ library and the R package code, or if you can't get a binary version of the latest C++ library elsewhere, you'll need to build it from source. This section discusses how to set up a C++ build configured to work with the R package. For more general resources, see the [Arrow C++ developer guide](https://arrow.apache.org/docs/developers/cpp/building.html).

> There are five major steps to the process — the first four are relevant to all Arrow developers, and the last one is specific to developers making changes to the R package.

Reviewer: This is a very minor point, so we might want to ignore it totally / gloss over it here, but technically "the first four are relevant to all Arrow developers" isn't _quite_ true. There are (small) differences in building the cpp code between the languages that use it (for example the flags recommended [for python development](https://arrow.apache.org/docs/developers/python.html#build-and-test) are slightly different — we could in principle recommend that folks use a union of the flags and then they would be able to do either, though compilation would be longer, so I'm not sure we want to do that). Additionally, other languages, like Rust, don't rely on the cpp code at all AFAIK, so for _those_ arrow developers this isn't super relevant.

All of that said, it might be best to leave it as is, or say something like "are similar to steps that Arrow developers working on other languages use" and leave it be. The details above in this comment are definitely *not* relevant for someone building Arrow for R development for the first time

Author: Thanks for highlighting this, I'll come up with a different way of phrasing this to be more accurate!

Author: I've just deleted it as it doesn't really add anything.

> ### Step 1 - Install dependencies

Reviewer: Did you intend to get rid of `{.tabset}` here? It's a bit funky how it works (and is very much non-standard/hacked on to pkgdown), but it gives us a nice tabbed interface for interacting with the setup steps (e.g. https://arrow.apache.org/docs/r/articles/developing.html#install-dependencies).

Author: Yeah, my intention was to make it really easy to navigate through the page, using the table of contents that appears at the side, and the heading levels, to signpost what bits applies to what.

I could add it back in though - will see how including it affects the contents list.

Reviewer: If the goal is to keep the tab titles in the table of contents, it's possible we could do that by changing some of the javascript that [we manipulate to make this whole thing work](https://github.com/apache/arrow/blob/master/r/pkgdown/extra.js). Something funny that we do there is that we load [the javascript for sticky tabs](https://cdn.jsdelivr.net/gh/rstudio/rmarkdown@47d837d3d9cd5e8e212b05767454f058db7d2789/inst/rmd/h/navigation-1.1/tabsets.js), and then alter some of the functions to do the things we need them to do in-place.

sudo apt install -y cmake libcurl4-openssl-dev libssl-dev
```

#### Windows

Currently, the R package cannot be made to work with a locally-built Arrow C++ library. This will be resolved in a future release.

> ### Step 2 - Configure the Arrow build

Reviewer: Same note about the `{.tabset}` here as well. If we do want to get rid of them completely, that's totally fine, but if we do get rid of the tabs, I think we should pick either user directory install or system install and only document one of those. Having both of these sections run one after the other on the the pkgdown site (which I assume is the place a large portion of people reading this read it from) is more confusing than it is helpful. Hiding one or the other with tabs makes it less distracting to document both cause one can totally ignore one/the other.

_Note_ `cmake` is particularly sensitive to whitespacing, if you see errors, check that you don't have any errant whitespace.

### Step 3 - Building libarrow

You can add `-j#` between `make` and `install` here too to speed up compilation by running in parallel (where `#` is the number of cores you have available).

```{bash, save=run & !(sys_install & ubuntu)}
make -j8 install
```

### Step 4 - Build the Arrow R package

Once you've built libarrow, you can install the R package and its
dependencies, along with additional dev dependencies, from the git
checkout:

```{bash, save=run}
popd # To go back to the root directory of the project, from cpp/build
pushd r
R -e 'install.packages("remotes"); remotes::install_deps(dependencies = TRUE)'
R CMD INSTALL .
```

#### Compilation flags

If you need to set any compilation flags while building the C++
extensions, you can use the `ARROW_R_CXXFLAGS` environment variable. For
example, if you are using `perf` to profile the R extensions, you may
need to set

```bash
export ARROW_R_CXXFLAGS=-fno-omit-frame-pointer
```

#### Recompiling the C++ code

> With the setup described here, you should not need to rebuild the Arrow library or even the C++ source in the R package as you iterated and work on the R package. The only time those should need to be rebuilt is if you have changed the C++ in the R package (and even then, `R CMD INSTALL .` should only need to recompile the files that have changed) _or_ if the libarrow C++ has changed and there is a mismatch between libarrow and the R package. If you find yourself rebuilding either or both each time you install the package or run tests, something is probably wrong with your set up.

Reviewer: ```suggestion
With the setup described here, you should not need to rebuild the Arrow library or even the C++ source in the R package as you iterate and work on the R package. The only time those should need to be rebuilt is if you have changed the C++ in the R package (and even then, `R CMD INSTALL .` should only need to recompile the files that have changed) _or_ if the libarrow C++ has changed and there is a mismatch between libarrow and the R package. If you find yourself rebuilding either or both each time you install the package or run tests, something is probably wrong with your set up.
```

Other flags that may be useful:

> * `-DBoost_SOURCE=BUNDLED` and `-DThrift_SOURCE=bundled`, for example, or any other dependency `*_SOURCE`, if you have a system version of a C++ dependency that doesn't work correctly with Arrow. This tells the build to compile its own version of the dependency from source.

Reviewer: ```suggestion
* `-DBoost_SOURCE=BUNDLED` and `-DThrift_SOURCE=BUNDLED`, for example, or any other dependency `*_SOURCE`, if you have a system version of a C++ dependency that doesn't work correctly with Arrow. This tells the build to compile its own version of the dependency from source.
```

I believe these are both not case sensitive (though we should probably check). This is something I noted when looking at this page a few days ago but didn't have a change to come and fix it separately.

devtools::document()

# To preview the documentation website
pkgdown::build_site(preview=TRUE)
```

## Styling and linting

### R code

The R code in the package follows [the tidyverse style](https://style.tidyverse.org/). On PR submission (and on pushes) our CI will run linting and will flag possible errors on the pull request with annotations.

To run the [lintr](https://github.com/jimhester/lintr) locally, install the lintr package (note, we currently use a fork that includes fixes not yet accepted upstream, see how lintr is being installed in the file `ci/docker/linux-apt-lint.dockerfile` for the current status) and then run 

```{r}
lintr::lint_package("arrow/r")
```

You can automatically change the formatting of the code in the package using the [styler](https://styler.r-lib.org/) package. There are two ways to do this:

1. Use the comment bot to do this automatically with the command `@github-actions autotune` on a PR, and commit it back to the branch.

2. Run the styler locally either via Makefile commands:

```bash
make style # (for only the files changed)
make style-all # (for all files)
```

or in R:

```{r}
# note the two excluded files which should not be styled

> styler::style_pkg(exclude_files = c("tests/testthat/latin1.R", "data-raw/codegen.R"))

Reviewer: ```suggestion
styler::style_pkg(exclude_files = c("tests/testthat/latin1.R", "data-raw/codegen.R"))

```

To appease the lintr, a bit silly.

* All tests are skipped on Linux if the package builds without the C++ libarrow.
  To make the build fail if libarrow is not available (as in, to test that
  the C++ build was successful), set `TEST_R_WITH_ARROW=true`
  
* Some tests are disabled unless `ARROW_R_DEV=true`

* Tests that require allocating >2GB of memory to test Large types are disabled
  unless `ARROW_LARGE_MEMORY_TESTS=true`
  
* Integration tests against a real S3 bucket are disabled unless credentials
  are set in `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`; these are available
  on request
  
* S3 tests using [MinIO](https://min.io/) locally are enabled if the
  `minio server` process is found running. If you're running MinIO with custom
  settings, you can set `MINIO_ACCESS_KEY`, `MINIO_SECRET_KEY`, and
  `MINIO_PORT` to override the defaults.

## Running checks

You can run package checks by using `devtools::check()` and check test coverage with `covr::package_coverage()`.

```r
# All package checks
devtools::check()

# See test coverage statistics
covr::report()
covr::package_coverage()

> 

Reviewer: ```suggestion
```

```

For full package validation, you can run the following commands from a terminal.

```
R CMD build .
R CMD check arrow_*.tar.gz --as-cran
```

## Running additional CI checks

On a pull request, there are some actions you can trigger by commenting on the PR. We have additional CI checks that run nightly and can be requested on demand using an internal tool called [crosssbow](https://arrow.apache.org/docs/developers/crossbow.html). A few important GitHub comment commands include:

#### Run all extended R CI tasks
`@github-actions crossbow submit -g r`

This runs each of the R-related CI tasks.

#### Run a specific task
`@github-actions crossbow submit {task-name}`

See the `r:` group definition near the beginning of the [crossbow configuration](https://github.com/apache/arrow/blob/master/dev/tasks/tasks.yml) for a list of glob expression patterns that match names of items in the `tasks:` list below it.

#### Run linting and documentation building tasks

`@github-actions autotune`

> This will run and fix lint C++ linting errors, run R documentation (among other cleanup tasks), and commit the resulting updates to the branch.

Reviewer: ```suggestion
This will run and fix lint C++ linting errors, run R documentation (among other cleanup tasks), run styler on any changed R code, and commit the resulting updates to the branch.
```

only necessary if you make changes to the C++ library source; you do not
need to manually purge object files if you are only editing R or C++
code inside `r/`.

## Arrow library-R package mismatches

If the Arrow library and the R package have diverged, you will see errors like:

```
Error: package or namespace load failed for ‘arrow' in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/Library/Frameworks/R.framework/Versions/4.0/Resources/library/00LOCK-r/00new/arrow/libs/arrow.so':
  dlopen(/Library/Frameworks/R.framework/Versions/4.0/Resources/library/00LOCK-r/00new/arrow/libs/arrow.so, 6): Symbol not found: __ZN5arrow2io16RandomAccessFile9ReadAsyncERKNS0_9IOContextExx
  Referenced from: /Library/Frameworks/R.framework/Versions/4.0/Resources/library/00LOCK-r/00new/arrow/libs/arrow.so
  Expected in: flat namespace
 in /Library/Frameworks/R.framework/Versions/4.0/Resources/library/00LOCK-r/00new/arrow/libs/arrow.so
Error: loading failed
Execution halted
ERROR: loading failed
```

To resolve this, try rebuilding the Arrow library from [Building Arrow above](#step-3-building-arrow).

## Multiple versions of Arrow library

> If rebuilding the Arrow library doesn't work and you are [installing from a user-level directory](#configure-for-installing-to-a-user-directory) and you already have a previous installation of libarrow in a system directory or you get you may get errors like the following when you install the R package:

Reviewer: ```suggestion
If rebuilding the Arrow library doesn't work and you are [installing from a user-level directory](#configure-for-installing-to-a-user-directory) and you already have a previous installation of libarrow in a system directory or you get errors like the following when you install the R package:
```

I'm not totally sure this is what you intended, but I think it is?

Author: No idea, I think I just moved this section from elsewhere rather than rewriting it and had just skim-read it so didn't catch this. I'll rewrite the whole sentence to break it down a little.

If the package fails to install/load with an error like this:

```
  ** testing if installed package can be loaded from temporary location
  Error: package or namespace load failed for 'arrow' in dyn.load(file, DLLpath = DLLpath, ...):
  unable to load shared object '/Users/you/R/00LOCK-r/00new/arrow/libs/arrow.so':
  dlopen(/Users/you/R/00LOCK-r/00new/arrow/libs/arrow.so, 6): Library not loaded: @rpath/libarrow.14.dylib
```

ensure that `-DARROW_INSTALL_NAME_RPATH=OFF` was passed (this is important on
macOS to prevent problems at link time and is a no-op on other platforms).
Alternatively, try setting the environment variable `R_LD_LIBRARY_PATH` to
wherever Arrow C++ was put in `make install`, e.g. `export
R_LD_LIBRARY_PATH=/usr/local/lib`, and retry installing the R package.

When installing from source, if the R and C++ library versions do not
match, installation may fail. If you've previously installed the
libraries and want to upgrade the R package, you'll need to update the
Arrow C++ library first.

For any other build/configuration challenges, see the [C++ developer
guide](https://arrow.apache.org/docs/developers/cpp/building.html).

## Other installation issues

> There are a number of scripts that are triggered when the arrow R package is installed. For package users who are not interacting with the underlying code, these should all just work without configuration and pull in the most complete pieces (e.g. official binaries that we host). However, knowing about these scripts can help package developers troubleshoot if things go wrong in them or things go wrong in an install.  See [the installation vignette](./install.html) for more information.

Reviewer: Could/should we link to the `How dependencies are resolved` section in the install vignette here?

Author: Sounds good
